### PR TITLE
Detach clouddrives while IO inprogress and verify pool can be back after maintenance cycle

### PR DESCRIPTION
**Pull request by Santhoshairody**
**What this PR does / why we need it**:
Detach clouddrives while IO inprogress and verify pool can be back after maintenance cycle

**Which issue(s) this PR fixes** (optional)
Closes #
PTX-16686
**Special notes for your reviewer**:
Pending tasks:
1. Corecheck and delete core if PX crashed during drives detachment.
2. Integrity check should be done.

### CODE REVIEW

File: drivers/node/vsphere/vsphere.go [hunk 1]
```diff
@@ -248,6 +332,139 @@ func (v *vsphere) connect() error {
 	return nil
 }
 
+// DetachDisk vdisks from node.
+func (v *vsphere) DetachDrivesFromVM(stc *corev1.StorageCluster, nodeName string) error {
+	configData, err := GetCloudDriveConfigmapData(stc)
+	//log.InfoD("Below is the cloud drive config data")
+	//log.InfoD("Config Data %v", configData)
```
Comment (lsrinivas-pure) [inline drivers/node/vsphere/vsphere.go:339]: delete if not required

File: drivers/node/vsphere/vsphere.go [hunk 2]
```diff
@@ -248,6 +332,139 @@ func (v *vsphere) connect() error {
...
+	return diskPath == volPath
+}
+
+//Get virtual disk path.
+//TODO need to filter only of type: DrivePaths
+func GetDiskPaths(driveset DriveSet) []string {
+	diskPaths := []string{}
+	for vmdkPath, configs := range driveset.Configs {
+		//TODO need to change later
+		log.InfoD("PX type %s ", configs.PXType)
```
Comment (lsrinivas-pure) [inline drivers/node/vsphere/vsphere.go:428]: use infoF if the log is not needed to push to aetos

File: tests/testTriggers.go [hunk 3]
```diff
@@ -813,6 +814,134 @@ func TriggerDeployNewApps(contexts *[]*scheduler.Context, recordChan *chan *Even
...
+	stepLog := "Detach cloud drives from Storage node"
+	Step(stepLog, func() {
+		var err error
+		var namespace string
+		log.InfoD(stepLog)
+		Step(stepLog,
+			func() {
+				log.InfoD(stepLog)
+				if namespace, err = Inst().S.GetPortworxNamespace(); err != nil {
+					log.InfoD("Namespace not available")
```
Comment (lsrinivas-pure) [inline tests/testTriggers.go:844]: if you do not want to continue after error use return

File: tests/testTriggers.go [hunk 4]
```diff
@@ -813,6 +814,134 @@ func TriggerDeployNewApps(contexts *[]*scheduler.Context, recordChan *chan *Even
...
+				pxOperator := operator.Instance()
+				stcList, err := pxOperator.ListStorageClusters(namespace)
+				if err != nil {
+					log.InfoD("Storage cluster not available")
+					UpdateOutcome(event, err)
+				}
+				log.InfoD("Stc list %v ",stcList)
+				storageNodes:= node.GetStorageNodes()
+				stc, err := pxOperator.GetStorageCluster(stcList.Items[0].Name, stcList.Items[0].Namespace)
+				if err != nil {
```
Comment (lsrinivas-pure) [inline tests/testTriggers.go:857]: if is not required UpdateOutcome has that check

File: tests/testTriggers.go [hunk 5]
```diff
@@ -813,6 +814,134 @@ func TriggerDeployNewApps(contexts *[]*scheduler.Context, recordChan *chan *Even
...
+				status, _ = Inst().V.GetNodeStatus(storageNodes[0])
+				log.InfoD("Node Status after exit maintenance mode %v , %s", status, storageNodes[0].Name)
+				if status.String() != apios.Status_STATUS_OK.String(){
+					UpdateOutcome(event, fmt.Errorf("Node %s: Expected: %v Actual: %v", storageNodes[0].Name,
+					apios.Status_STATUS_MAINTENANCE, status))
+				}
+				//There are chances that storageless node can convert storage node by attaching these drives. Hence making sure
+				//This node id is up.
+				var selectedStorageNode *node.Node
+				storageNodes = node.GetStorageNodes()
```
Comment (lsrinivas-pure) [inline tests/testTriggers.go:909]: After detaching and doing doing node maintenance cycle we need to refresh the driver endpoints.
here node.GetStorageNodes() will have the same data as before tests start as it is not refreshed

File: tests/testTriggers.go [hunk 6]
```diff
@@ -813,6 +815,132 @@ func TriggerDeployNewApps(contexts *[]*scheduler.Context, recordChan *chan *Even
...
+					ctx.SkipVolumeValidation = false
+					errorChan:= make(chan error, errorChannelSize)
+					ValidateContext(ctx, &errorChan)
+					for err := range errorChan {
+						UpdateOutcome(event, err)
+					}
+				}
+				err = ValidateDataIntegrity(contexts)
+				UpdateOutcome(event, err)
+				dash.VerifyFatal(err==nil, true, "Data integrity check ?")
```
Comment (lsrinivas-pure) [inline tests/testTriggers.go:938]: dont use VerifyFatal in longevity, if fails, it stops the longevity

File: tests/testTriggers.go [hunk 7]
```diff
@@ -813,6 +815,132 @@ func TriggerDeployNewApps(contexts *[]*scheduler.Context, recordChan *chan *Even
...
+						break
+					}
+				}
+				if selectedStorageNode != nil{
+					status, err = Inst().V.GetNodeStatus(*selectedStorageNode)
+					if err != nil {
+						UpdateOutcome(event, err)
+					}else{
+						log.Infof("Node Status  %v node: %s", status, (*selectedStorageNode).SchedulerNodeName)
+						if status.String() != apios.Status_STATUS_OK.String(){
```
Comment (lsrinivas-pure) [inline tests/testTriggers.go:919]: if node id is same doest it mean that same driveset is added?

Comment (Santhoshairody) [reply tests/testTriggers.go:919]: Yes. node id is nothing but driveset id, if storageless node becomes storagenode then nodeId is moved to new storage node.

File: drivers/node/vsphere/vsphere.go [hunk 8]
```diff
@@ -43,6 +48,85 @@ const (
...
+	InstanceID string
+	// Zone defines the zone in which the node exists
+	Zone string
+	// State state of the drive set from the well defined states
+	State DriveSetState
+	// Labels associated with this drive set
+	Labels *map[string]string `json:"labels"`
+}
+
+// DriveSetState indicates the state of the DriveSet
```
Comment (Rohit-PX) [inline drivers/node/vsphere/vsphere.go:77]: Can just define it as string?

File: drivers/node/vsphere/vsphere.go [hunk 9]
```diff
@@ -43,6 +48,85 @@ const (
...
+type DriveSetState string
```
Comment (Rohit-PX) [inline drivers/node/vsphere/vsphere.go:78]: Same as above.

File: drivers/node/vsphere/vsphere.go [hunk 10]
```diff
@@ -43,6 +48,85 @@ const (
...
+
+// PXDriveType defines a set of drive types based on how they are used by PX
+type PXDriveType string
```
Comment (Rohit-PX) [inline drivers/node/vsphere/vsphere.go:81]: Same as above.

File: drivers/node/vsphere/vsphere.go [hunk 11]
```diff
@@ -248,6 +332,137 @@ func (v *vsphere) connect() error {
 	return nil
 }
 
+// DetachDisk vdisks from node.
+func (v *vsphere) DetachDrivesFromVM(stc *corev1.StorageCluster, nodeName string) error {
+	configData, err := GetCloudDriveConfigmapData(stc)
```
Comment (Rohit-PX) [inline drivers/node/vsphere/vsphere.go:337]: Check for `if err != nil`

File: drivers/node/vsphere/vsphere.go [hunk 12]
```diff
@@ -248,6 +332,137 @@ func (v *vsphere) connect() error {
...
+	for _, nodeConfigData := range configData {
+		if nodeName == nodeConfigData.SchedulerNodeName {
+			allDiskPaths := GetDiskPaths(nodeConfigData)
+			instanceId := nodeConfigData.InstanceID
+			for i := 0; i < len(allDiskPaths); i++ {
+				log.Infof("Diskpath for %v is %v and instance id is %v", nodeConfigData.NodeID, allDiskPaths[i], instanceId)
+				err = v.DetachDisk(instanceId, allDiskPaths[i])
+				if err != nil {
+					//log.InfoD("Detach drives from the node failed %v", err)
+					err = fmt.Errorf("%w; %s", err, nodeName)
```
Comment (Rohit-PX) [inline drivers/node/vsphere/vsphere.go:348]:
```suggestion
					err = fmt.Errorf("Detaching disk: %s on node %s failed: %w", allDiskPaths[i],nodeName, err)
```

File: drivers/node/vsphere/vsphere.go [hunk 13]
```diff
@@ -248,6 +332,137 @@ func (v *vsphere) connect() error {
...
+			vmMo = vm
+			log.Infof("VM found %v", vm)
+			break
+		}
+	}
+	//Error if not found
+	if vmMo == nil {
+		return fmt.Errorf("Virtual machine not found")
+	}
+	log.Infof("Virtual machine %v ", vmMo)
```
Comment (Rohit-PX) [inline drivers/node/vsphere/vsphere.go:385]: Can delete this line. We don't need to print the full VM object.

File: drivers/node/vsphere/vsphere.go [hunk 14]
```diff
@@ -248,6 +332,137 @@ func (v *vsphere) connect() error {
...
+	//Remove device and detach VM
+	var deviceList object.VirtualDeviceList
```
Comment (Rohit-PX) [inline drivers/node/vsphere/vsphere.go:387]: We don't need to define it initially.

File: drivers/node/vsphere/vsphere.go [hunk 15]
```diff
@@ -248,6 +332,137 @@ func (v *vsphere) connect() error {
...
+	var selectedDevice types.BaseVirtualDevice
+	deviceList, err = vmMo.Device(v.ctx)
```
Comment (Rohit-PX) [inline drivers/node/vsphere/vsphere.go:389]:
```suggestion
	deviceList, err := vmMo.Device(v.ctx)
```

**Review by Rohit-PX — CHANGES_REQUESTED** [review]
Looks good otherwise.

**Review by Rohit-PX — APPROVED** [review]

**Pull request merged**